### ISSUE TITLE
LSP: undoing after vim.lsp.buf.format() changes cursor position

### ISSUE THREAD

**Issue by pantosaur** [neovim/neovim#24297]
### Problem

I tried with null-ls, and lsp formatters, and the behavior is that whenever calling vim.lsp.buf.format(), the cursor remains in position, which is what I would expect.

However, when undoing the formatting changes, I would expect the same behavior, seeing the changes revert back on the exact same lines displayed in the window.
But instead, the cursor jumps to the top of the window, more specifically to the lowest line number where a change was undone.

This is annoying because it makes it difficult to undo/redo formatting a couple of times to view the changes happening in a specific area.

### Steps to reproduce

open a file for which you have a formatter installed
make sure the file is not formatted in multiple areas
call vim.lsp.buf.format()
undo

### Expected behavior

the cursor and window position remains the same

### Neovim version (nvim -v)

NVIM v0.9.1 Build type: Release LuaJIT 2.1.0-beta3

### Vim (not Nvim) behaves the same?

n/a

### Operating system/version

Linux 6.4.1-artix1-1

### Terminal name/version

alacritty 0.12.2

### $TERM environment variable

alacritty

### Installation

artix repo

**Comment by justinmk** [comment]
> But instead, the cursor jumps to the top of the window, more specifically to the lowest line number where a change was undone.

This is just how undo works in vim. If other lsp formatters are behaving differently, maybe we can borrow their approach if it's not complicated.

We should also use `:keepmarks` to avoid setting `'[` and `']` to the start/end of the file. https://github.com/neovim/neovim/issues/8807

Related:

- #22253
- https://github.com/neovim/neovim/issues/21783

**Comment by duarm** [comment]
Let me know if there's any workarounds, something like a undo.stop_recording() and undo.continue_recording(), I like to trim trailing white spaces  on save, but this messes my undo history, the same solution could be used I suspect

**Comment by faergeek** [comment]
>This is just how undo works in vim. If other lsp formatters are behaving differently, maybe we can borrow their approach if it's not complicated.

@justinmk it seems like not moving cursors at all after undoing such changes would make sense, though I see that at least current window cursor is moved intentionally and explicitly to the start of the first changed line in such cases:

https://github.com/neovim/neovim/blob/aab06edc63c3dfbd99e0c803c6834bf771318c42/src/nvim/undo.c#L2312-L2327

I tried simply commenting out this code and then cursor is not being moved after undo. This code is from the original vim, commenting it out didn't seem to break any old tests and I couldn't track down where this was introduced to understand why exactly it is there. A couple of current functional tests rely on this behavior though. And I believe simply removing it is not an option, so how could we approach this?

It could probably be solved by either introducing some flag telling that cursor should not be moved or writing invalid cursor position, which would indicate the same. And then use that from `nvim_buf_set_text`, which is used for lsp formatting. Though both of these solutions seem quite ugly. Are there any other solutions I might be missing?

**Comment by justinmk** [comment]
> couldn't track down where this was introduced to understand why exactly it is there

Likely matches POSIX vi.

> And I believe simply removing it is not an option, so how could we approach this?

If it is unlikely that scripts depend on the behavior then we could just remove it, if the behavior is genuinely an ergonomics upgrade. But see https://github.com/neovim/neovim/issues/5989 for more discussion.

Unless I'm mistaken, this is a duplicate of https://github.com/neovim/neovim/issues/5989 , so closing.

**Comment by faergeek** [comment]
@justinmk I don't think this is actually a duplicate, because behavior mentioned here seems to be about lines only, not columns, at least from what I see in the code and from basic experiments and tests. But example provided in #5989 is about edits on a single line, which only involves movements over columns.

**Comment by justinmk** [comment]
I see https://github.com/neovim/neovim/issues/5989 as covering the general topic, the particular example it has is just an example. But we can keep this open.

**Comment by zeertzjq** [comment]
This was introduced before Vim had tests: https://github.com/vim/vim-history/commit/0206c5b14035929b9392a09985ab4b9fdef94003

**Comment by justinmk** [comment]
Thanks, that has a useful hint: whenever the behavior is "try to ..." that means scripts can't depend on it, thus we can change it.

> Solution:   Try to move the cursor back to where it was or to the first line that actually changed.

**Comment by faergeek** [comment]
> Let me know if there's any workarounds, something like a undo.stop_recording() and undo.continue_recording(), I like to trim trailing white spaces on save, but doing so messes up my undo history

@duarm seems like `:undojoin` could help, though there's a catch, see https://github.com/sbdchd/neoformat/issues/386

**Comment by justinmk** [comment]
>  seems like `:undojoin` could help, though there's a catch, see [sbdchd/neoformat#386](https://github.com/sbdchd/neoformat/issues/386)

What about `:help i_CTRL-G_u` + `:help i_CTRL-G_U` ? Could those be used to fashion the desired undo positions?

**Comment by faergeek** [comment]
> What about `:help i_CTRL-G_u` + `:help i_CTRL-G_U` ? Could those be used to fashion the desired undo positions?

I don't see how those mappings would help in that cases. `:help i_CTRL-G_u` starts a new undo block, which seems to be the opposite of what was requested. And IIUC, `:help i_CTRL-G_U` allows to move cursor along the line without closing current undo block, which in theory could help if trimming is implemented by redefining mappings which move cursor up or down in insert mode to do trimming as well, but that seems complicated, so I doubt that's the case here.

**Comment by faergeek** [comment]
https://github.com/neovim/neovim/issues/24297#issuecomment-1715002533

> it seems like not moving cursors at all after undoing such changes would make sense

Update. Moving cursor by how line count above cursor changed after undo still makes sense, just without pulling it exactly to the lowest line number. I tried not moving it at all in this PR https://github.com/neovim/neovim/pull/25119. This behavior seems to be much worse than the current one in cases where line count above cursor changes quite significantly. I tried it and I could end up inside a totally different function after undoing formatting because of that.

**Comment by Sam-programs** [comment]
this seems to work and idk why
```lua
keymap("n", "<leader>f", function()
   local lines = vim.api.nvim_buf_get_lines(0, 0, -1, false)
   vim.api.nvim_buf_set_lines(0, 0, -1, false, lines)
   vim.lsp.buf.format()
end)
```

**Comment by ceuk** [comment]
Just want to add to this that C-i and C-o don't seem to work as expected after this occurs either. (which makes it even harder to get back to where you were)

**Comment by dhensen** [comment]
Had this issue for months, only suffered from this when editing javascript though, not while doing python, so I figured it was null-ls or prettier. Then googling, but this was a little harder to find because I was looking for null-ls/prettier/JSX/JS related search terms.

Why don't I suffer from this when doing python? Is the undojoin story above really the full story?

**Comment by faergeek** [comment]
@dhensen it depends on where the changes have been applied after formatting. It may occur more often for javascript if e.g. you're doing an auto-import which inserts something like `import fs from "fs"` and formatter changes double quotes to single quotes, or may be adjusts spaces inside curly or something like that. Or in cases where formatter sends the whole buffer content as a replacement. In my case it was often caused by `esilnt-plugin-simple-import-sort`. I guess python simply doesn't have anything to format in imports, but I may be wrong, I don't code in python.

In any case cursor will basically just jump to the top-most change after undoing formatting done by `vim.buf.lsp.format`.

**Comment by dhensen** [comment]
@faergeek this is helpfull. On closer look it puts me back at the top of my function. When developing react and having a single component per file this is most often the only function, so it will appear as the top of the file, but it's actually the first line of the body of my function so:

```
import 1;
import 2;

export default function Foobar() {
   console.log('here');
   // 300 LoC
...
   // my cursor was here, I want to undo. and then I'm suddenly at console.log and I have to press u one more time to actually undo.
}

So my next step is going to see how prettier or null-ls/none-ls format the code. Maybe the function body is being replaced or line by line... I have no idea how formatters work. Thanks for the info

**Comment by lehmacdj** [comment]
Adding a minimal repro with `nvim --clean`. I'm able to reproduce this using multiple LSPs (at minimum: ruff, HLS, and a custom one I use).

I'm able to confirm that the cursor position is only updated after undo when the location of changes made by autoformatting occur outside of the visible.

https://github.com/lehmacdj/nvim-autoformat-undo-cursor-jump-minimal-setup

**Comment by 12joan** [comment]
Here's the workaround I'm using when formatting code on save.

```lua
-- Formatting code produces noise in the undo history,
-- especially when two formatters are in conflict. This means
-- that undoing can jump the cursor to unexpected locations and
-- redo history can be lost.
--
-- We can mitigate this by calling vim.cmd.undojoin prior to
-- formatting. However, this can throw an error when calling it
-- after an undo. We catch this error with pcall (a built-in Lua
-- function) and refrain from formatting if this occurs.
--
-- See:
-- - https://github.com/neovim/neovim/issues/24297
-- - https://www.reddit.com/r/neovim/comments/13xfltc/is_it_possible_to_ignore_last_entry_from_undo/
if pcall(vim.cmd.undojoin) then
  vim.lsp.buf.format({
    bufnr = args.buf,
    id = client.id,
    timeout_ms = 1000,
  })
end
```
